clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp

```diff
@@ -752,13 +787,17 @@ class ScalarExprEmitter : public StmtVisitor<ScalarExprEmitter, mlir::Value> {
                               QualType DstType, mlir::Type SrcTy,
                               mlir::Type DstTy, ScalarConversionOpts Opts);
 
-  BinOpInfo buildBinOps(const BinaryOperator *E) {
+  BinOpInfo buildBinOps(const BinaryOperator *E,
+                        QualType PromotionType = QualType()) {
     BinOpInfo Result;
-    Result.LHS = Visit(E->getLHS());
-    Result.RHS = Visit(E->getRHS());
-    Result.FullType = E->getType();
-    Result.CompType = E->getType();
-    if (auto VecType = dyn_cast_or_null<VectorType>(E->getType())) {
+    Result.LHS = CGF.buildPromotedScalarExpr(E->getLHS(), PromotionType);
+    Result.RHS = CGF.buildPromotedScalarExpr(E->getRHS(), PromotionType);
+    if (!PromotionType.isNull())
+      Result.FullType = PromotionType;
+    else
+      Result.FullType = E->getType();
+    Result.CompType = Result.FullType;
+    if (const auto *VecType = dyn_cast_or_null<VectorType>(Result.FullType)) {
       Result.CompType = VecType->getElementType();
     }
     Result.Opcode = E->getOpcode();
```

Comment:
While here, can you plz add an unimplented for `getFPFeaturesInEffect` below?

Reply:
Updated.